File: programers/find_prime.py
# 소수 찾기.  2022-12-13


### Python

def solution(n):
    p_set = set()
    ans = 0
    for i in range(2, n + 1):
        if i in p_set:
            continue
        
        ans += 1
        temp = i
        while temp <= n:
            p_set.add(temp)
            temp += i
    
    return ans


### JS

# function solution(n) {
    
#     const pSet = new Set()
#     let ans = 1
    
#     pSet.add(2)
#     for (let i = 3; i <= n; i += 2) {
#         if (pSet.has(i)) continue
        
#         ans += 1
#         temp = i
#         while (temp <= n) {
#             pSet.add(temp)
#             temp += i
#         }
#     }
    
#     return ans
# }


### Java

# class Solution {
#     public int solution(int n) {
        
#         int ans = 1;
#         for (int i = 3; i <= n; i = i + 2) {
#            if (primeTest(i)) ans += 1;
#         }
        
#         return ans;
#     }
    
#     public boolean primeTest(int n) {
#         for (int i = 2; i <= (int) Math.sqrt(n); i++) {
#             if (n % i == 0) return false;
#         }
#         return true;
#     }
# }

File: programers/find_prime.cpp
// 소수 찾기.  2023-11-09


#include <string>

using namespace std;

bool memo[100000000];
int counts[10];
int N;
int max_n;
int ans;

void make_nums(int k, int num)
{
    if (!memo[num]) ++ans;
    memo[num] = true;
    max_n = max(max_n, num);
    
    for (int i = 0; i < 10; ++i)
    {
        if (counts[i] != 0)
        {
            counts[i]--;
            make_nums(k + 1, num * 10 + i);
            counts[i]++;
        }
    }
}

int solution(string numbers)
{
    N = numbers.size();
    for (int i = 0; i < N; ++i)
    {
        counts[numbers[i] - '0']++;
    }
    
    make_nums(0, 0);
    if (memo[0]) --ans;
    if (memo[1]) --ans;
    
    int n = 2;
    while (n * n <= max_n)
    {
        int temp = n * 2;
        while (temp <= max_n)
        {
            if (memo[temp]) --ans, memo[temp] = false;
            temp += n;
        }
        ++n;
    }
    
    return ans;
}
